### PR TITLE
Use zuilnt for linting.

### PR DESCRIPTION
**Pull request by amanagr**
I have mostly moved the code around to make it compatible with zulint style. 
One important change I would like to mention is, now `./tools/lint` also runs `./tools/run-mypy`. I can easily disable it too.

### CODE REVIEW

File: tools/custom_check.py [hunk 1]
```diff
@@ -1,126 +1,37 @@
...
-         'description': 'Fix trailing whitespace'},
-        {'pattern': '^#+[A-Za-z0-9]',
-         'strip': '\n',
-         'description': 'Missing space after # in heading'},
-    ]  # type: RuleList
-    python_rules = cast(RuleList, [
+from typing import cast, Any, Dict, List, Tuple
+from zulint.custom_rules import RuleList
+
+Rule = List[Dict[str, Any]]
```
[the remark this answers is not in the document]
Comment (amanagr) [reply tools/custom_check.py]: Actually, it is present in zulint but due to python2 style type declaration in zulint, it is inside `if False:` statement and cannot be imported here.

File: tools/lint [hunk 2]
```diff
@@ -1,39 +1,57 @@
...
+        return 1 if failed else 0
+
+    @linter_config.lint
+    def pep8():
+        # type: () -> int
+        """Standard Python style linter on 50% of files (config: tools/linter_lib/pep8.py)"""
+        failed = check_pep8(by_lang['py'])
+        return 1 if failed else 0
+
+    linter_config.do_lint()
```
[the remark this answers is not in the document]
Comment (amanagr) [reply tools/lint]: It happens internally now inside the `do_lint()` function.

File: zulip_bots/zulip_bots/bots/google_translate/fixtures/test_invalid_api_key.json [hunk 3]
```diff
@@ -8,12 +8,12 @@
     }
   },
   "response": {
-    "error":{
```
[the remark this answers is not in the document]
Comment (amanagr) [reply zulip_bots/zulip_bots/bots/google_translate/fixtures/test_invalid_api_key.json]: Yeah, this was picked up zulint and should have been picked up by the old linter too. (I think this was a bug in it.) Okay, so I will move this commit to be the first one!

File: tools/custom_check.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andersk) [reply tools/custom_check.py]: We should change those to `from typing import TYPE_CHECKING; if TYPE_CHECKING:`.

File: tools/lint (hunk 2 above)
[the remark this answers is not in the document]
Comment (andersk) [reply tools/lint]: That suggests that `do_lint` should have type `() -> NoReturn`.

File: tools/custom_check.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanagr) [reply tools/custom_check.py]: Yeah!

**Comment by amanagr** [comment]
@andersk updated with requested changes!

File: requirements.txt [hunk 4]
```diff
@@ -7,3 +7,4 @@ pytest
 -e ./zulip
 -e ./zulip_bots
 -e ./zulip_botserver
+-e git+https://github.com/zulip/zulint@aaed679f1ad38b230090eadd3870b7682500f60c#egg=zulint==0.0.1
```
Comment (andersk) [inline requirements.txt]: `tools/provision` now warns: `WARNING: Requested zulint==0.0.1 from git+https://github.com/zulip/zulint@aaed679f1ad38b230090eadd3870b7682500f60c#egg=zulint==0.0.1 (from -r /home/anders/zulip/python-zulip-api/requirements.txt (line 10)), but installing version 1.0.0`.

(We have a bunch of these in zulip/zulip too. I was planning to clean them up soon.)

**Comment by timabbott** [comment]
@amanagr do you have a chance to update this?

File: requirements.txt (hunk 4 above)
Comment (amanagr) [reply requirements.txt]: Done!

**Comment by amanagr** [comment]
I am getting an error based on [this issue](https://github.com/PyCQA/pycodestyle/issues/728).
```
pep8      | /Users/amanagrawal/python-zulip-api/zulip-api-py3-venv/lib/python3.7/site-packages/pycodestyle.py:113: FutureWarning: Possible nested set at position 1
pep8      |   EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')
```

The solution is to upgrade pycodestyle to v2.4 or above. While this fixes the warning we get, it introduces around 100 new errors due addition of new rules to it. I will upgrade it perhaps in a different new PR.

**Comment by timabbott** [comment]
Sorry I missed your reply!  Looking now.

**Comment by timabbott** [comment]
This is great, merged, thanks @amanagr!  Sorry again that I lost track of this.

I think doing a new PR to upgrade our dependencies makes sense.

**Pull request merged**